### ISSUE TITLE
Make HiGHS installable via PyPI

### ISSUE THREAD

**Issue by l-kotzur** [ERGO-Code/HiGHS#925]
I would be great if one could call `pip install highs`. This would make the solver in general much easier installable and accessible.

@michaelbynum wrote already an appsi wrapper from Pyomo to HiGHS, as pointed in Pyomo/pyomo#2094. Unfortunately, this has stalled and has not been merged to the main branch. One of the main reasons is that HiGHS is not installable via a package manager. Therefore, it cannot be integrated into a proper CI-pipeline, I guess.

@mckib2 wrote wrappers for sckit and for scipy #156 and brought older versions to [PyPI](https://pypi.org/project/scikit-highs/). Would it be possible to make HiGHs as an own PyPI package such that Pyomo and others can more easily use it?

Unfortunately, I personally do not have any experience in packaging C++ libraries.

**Comment by mckib2** [comment]
> @mckib2 wrote wrappers for sckit and for scipy #156 and brought older versions to PyPI. Would it be possible to make HiGHs as an own PyPI package such that Pyomo and others can more easily use it?

The old scikit that you linked to was a proof-of-concept for integrating into the SciPy build system. All that work [has been integrated into SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html) and improved upon with the [MIP solver accessible](https://scipy.github.io/devdocs/reference/generated/scipy.optimize.milp.html) in the upcoming 1.9.0 release. Instead of the old buildutils build system, we now use meson. We expect to develop an object oriented interface in the coming months. 1.9.0 should also contain a very recent version of HiGHS with SciPy binaries available across many platforms and installable via pip. 

Any reason you can't use/build on SciPy's linprog function? I believe cvxpy is [taking this approach](https://www.google.com/url?sa=t&source=web&rct=j&url=https://github.com/cvxpy/cvxpy/issues/1721&ved=2ahUKEwivs5ufv5n5AhWoJ0QIHbbRC5YQFnoECAQQAQ&usg=AOvVaw2gSg9bXJkAxfYyrJhEKPFI) and you'll get support for Python specific build and usage issues

**Comment by jajhall** [comment]
> Any reason you can't use/build on SciPy's linprog function? I believe cvxpy is [taking this approach](https://www.google.com/url?sa=t&source=web&rct=j&url=https://github.com/cvxpy/cvxpy/issues/1721&ved=2ahUKEwivs5ufv5n5AhWoJ0QIHbbRC5YQFnoECAQQAQ&usg=AOvVaw2gSg9bXJkAxfYyrJhEKPFI) and you'll get support for Python specific build and usage issues

Thanks for the pointer on this @mckib2 We need HiGHS to be installable via a package manager for a number of projects. @michaelbynum and @galabovaa have managed this for MacOS and Linux, but Windows is proving to be very tricky. 

I don't know whether we can use/build on SciPy's linprog function, but @galabovaa is soon back from holiday and will have a view

**Comment by l-kotzur** [comment]
@jajhall Only supporting Linux and MacOS would be fine at least for me. Coin-CBC is only supported on Linux and MacOS as well atm, if I am not wrong.

**Comment by jajhall** [comment]
For the prototype "thin" Python wrapper `highspy` around the C++ class, `pip install highspy' works on two Ubuntu installations, and is not known to fail on any other Linux installation.

**Comment by l-kotzur** [comment]
@michaelbynum Would this be sufficient to integrate the appsi-Highs interface to pyomo?

**Comment by jajhall** [comment]
pip install highspy also works on (at least one) MacOS. Windows is WIP.

**Comment by stephane-caron** [comment]
Having ``highspy`` on PyPI is great! It helped me a lot in adding an interface for HiGHS in [qpsolvers](https://github.com/stephane-caron/qpsolvers#solvers).

Looking forward to the upcoming release (since currently 1.1.2.dev3 is pre-release) :smiley: 

Also, not part of this issue but likely a follow-up feature request, I'll be looking forward for the addition of ``get/setHotStart`` to the interface (keeping track in https://github.com/stephane-caron/qpsolvers/issues/94).

**Comment by jajhall** [comment]
> Also, not part of this issue but likely a follow-up feature request, I'll be looking forward for the addition of `get/setHotStart` to the interface (keeping track in [stephane-caron/qpsolvers#94](https://github.com/stephane-caron/qpsolvers/issues/94)).

Hot starting the QP solver is not possible at present, and might not be possible for a long time

**Comment by michaelbynum** [comment]
I could not find highspy on PyPI. I know pip install works locally, but is it on PyPI yet? If so, could anyone share a link?

@l-kotzur, yes - having highspy on PyPI would be sufficient to integrate the appsi-HiGHS interface into Pyomo.

**Comment by michaelbynum** [comment]
Thanks!

**Comment by jackbrucesimpson** [comment]
Hi, do you know when/if there were plans for the pip installed version to work on Windows? Thanks so much.

**Comment by jajhall** [comment]
We certainly "plan" to get pip install highspy to work on Windows, but achieving it is another matter!

**Comment by michaelbynum** [comment]
@l-kotzur, The Pyomo-Appsi interface to HiGHS is now in Pyomo main and will be in the next release.

**Comment by l-kotzur** [comment]
> @l-kotzur, The Pyomo-Appsi interface to HiGHS is now in Pyomo main and will be in the next release.

I have seen it. Those are great news, thanks. I am looking forward to test it.

**Comment by trigaut** [comment]
Hello @jajhall, what is the issue to pip install on windows? I would be glad to help if you think I could. We develop on windows where I work.

The [HiGHS Rust interface](https://github.com/rust-or/highs-sys) works on Windows when we install cmake and llvm. But I imagine it is more challenging with PyPI?

**Comment by jajhall** [comment]
> Hello @jajhall, what is the issue to pip install on windows? I would be glad to help if you think I could. We develop on windows where I work.

I don't know specifically, because it's being worked on by @galabovaa. Currently she's combining work on this with integrating HiGHS into another major solver package. Perhaps there's some specific issue that you could help her with, but I'll leave her to ask. Many thanks!

**Comment by galabovaa** [comment]
Hi @trigaut 

Currently on Windows, pip install seems to be generating a wheel, however, when I call "import highspy" from python I get a "DLL not found" error. 

I am not an expert on Windows and I have been struggling to get the paths correct.

Any help would be much appreciated. You could have a look at the win branch which contains some minor changes to the CMake files. 

Thanks, 
Ivet

**Comment by jajhall** [comment]
@trigaut: It's worth (re-?)emphasising that `pip install highspy` works on Linux and MacOS

**Comment by galabovaa** [comment]
I just managed to build a wheel for Windows 64bit cp39: https://pypi.org/project/highspy/

If you would please try to install from your windows machines and test if it works on your side that would be great!

**Comment by l-kotzur** [comment]
The example from the README is working for my case on Windows 10. Great support, thanks!

**Comment by jackbrucesimpson** [comment]
Version 1.1.2.dev3 was working on Linux but when I just tested out updating
to 1.4.0.dev0 it appears to fail.

[image: image.png]

**Comment by jajhall** [comment]
> Version 1.1.2.dev3 was working on Linux but when I just tested out updating to 1.4.0.dev0 it appears to fail.

 `1.4.0.dev0` doesn't have a recent branch from @galabovaa merged into it

**Comment by galabovaa** [comment]
Yesterday I uploaded highspy 1.4.0 only for Windows so trying to update on Linux would give users an error. I will update linux and macOS as well to 1.4.0 once we merge the latest changes into master.

**Comment by jackbrucesimpson** [comment]
Thanks so much!

**Comment by trigaut** [comment]
Just tested `pip install highspy` and to solve many problems modeled with pyomo and using the "appsi_highs" SolverFactory. It works like a charm! Thanks !

**Comment by jajhall** [comment]
@galabovaa I've tested it on my Windows laptop, and it works perfectly!

**Comment by michaelbynum** [comment]
Excellent news! I would also like to thank @galabovaa for all the effort putting together the PyPI releases. It made things much easier for Pyomo.

**Comment by julian-belina** [comment]
Is it planned to add a source distribution to the pypi release? This would simplify the creation of conda package.

**Comment by jajhall** [comment]
> Is it planned to add a source distribution to the pypi release? This would simplify the creation of conda package.

If it's not difficult, then I'll ask @galabovaa to add one

**Comment by rschwarz** [comment]
Another related question: I saw that a version v1.5.1 was released on GitHub some days ago, but the most recent version for `highspy` on PyPI is still `v1.5.0.dev0`. Does that mean that the `highspy` packages are created & published manually?

**Comment by jajhall** [comment]
The `highspy` packages are created & published manually. Automating this is WIP

**Comment by yasirroni** [comment]
`pip install highspy` is no longer works in Windws in my case, returning "DLL not found" error, again

```
Requirement already satisfied: highspy (1.5.0.dev0)
Requirement already satisfied: pyomo (from highspy) (6.5.0)
Requirement already satisfied: pybind11 (from highspy) (2.10.4)
Requirement already satisfied: numpy (from highspy) (1.23.5)
Requirement already satisfied: ply (from pyomo->highspy) (3.11)
```

Specify previous version seems to not works

```
pip uninstall highspy
WARNING: Skipping highspy as it is not installed.

pip install highspy==1.4.0.dev0
ERROR: Could not find a version that satisfies the requirement highspy==1.4.0.dev0 (from versions: 1.5.0.dev0)
ERROR: No matching distribution found for highspy==1.4.0.dev0
```

**Comment by jajhall** [comment]
Sorry about this. `highspy` is still WIP by @galabovaa, in particular its installation on Windows.

**Comment by ryanjoneil** [comment]
Are you all looking for some help with this? :smile_cat:

**Comment by jajhall** [comment]
If you've got someone experienced in building wheels for PyPI, then @galabovaa would welcome someone with whom to discuss issues

**Comment by galabovaa** [comment]
Some help would be much appreciated, thank you @ryanjoneil 

I will send you an email with some details about what the particular issues are, if you could point me to a feasible direction that would be great!

**Comment by galabovaa** [comment]
highspy is now on PyPi, with a pre-release v1.7.1.dev1 already up and v1.7.1 coming up very soon!

**Comment by julian-belina** [comment]
@galabovaa will v1.7.1 include an sdist file? This would simplify the creation of a cond package.

**Comment by julian-belina** [comment]
> @galabovaa will v1.7.1 include an sdist file? This would simplify the creation of a conda package.

**Comment by galabovaa** [comment]
Yes, it will include an sdist, apologies for the delay. If you wish to check whether the conda build will work, v 1.7.1.dev1 should already contain one: 

https://pypi.org/project/highspy/1.7.1.dev1/#files

**Comment by julian-belina** [comment]
Seems to work, thank you verry much!

**Comment by julian-belina** [comment]
Hey @Jeromeschmidt , unfortunately I was not able to make the conda forge build realy work. I am not sure whats causing the problem.

Maybe its a problem with Highs as discussed in this Issue https://github.com/ERGO-Code/HiGHS/issues/1119 . The Issue was fixed in the pre releaese 1.7.1.dev1, but there is has not been published as a stable release so far. I dont know how if there is a way to use the pre release to build conda-forge package from a pre release.

There is also the possibility that the recipe is not setup correctly. I am not too familiar with cmake. If you are, feel free to take a look a the recipe:
https://github.com/conda-forge/highs-feedstock/tree/main/recipe

If you have suggestions to improve the recipe, I can integrate them or make you a maintainer.
